### PR TITLE
[edk2-stable202411] MdeModulePkg/DxeIplPeim: Fix pagetable protection region in 5 leve paging

### PR DESCRIPTION
**Pull request by NingFengGit**
MedModulePkg/DxeIplPeim: Fix pagetable protection region in 5 leve paging

REF:https://bugzilla.tianocore.org/show_bug.cgi?id=4873

Currently the function does not cover the 5 level paging case. it will caused pagetable protection region to be set incorrectly. This patch do the enhancement and with the patch protection region has been set correctly.

CC: Ray Ni [ray.ni@intel.com](mailto:ray.ni@intel.com)

##Description
Currently the function does not cover the 5 level paging case. it will casued pagetable protection region set incorrectly. This patch do the enhancemant and with the patch protection region has been set correctly.

##How This Was Tested
we test on real platform with below condition:
5level pagetable +UPL, pagetable region been protected correctly.
4level pagetable+UPL, pagetable region been protected correctly.

5level pagetable +nonUPL pagetable region been protected correctly.
4level pagetable+nonUPL, pagetable region been protected correctly.

### CODE REVIEW

File: MdeModulePkg/Core/DxeIplPeim/X64/VirtualMemory.c [hunk 1]
```diff
@@ -485,14 +485,14 @@ Split1GPageTo2M (
...
+  @param[in] LevelofPaging     Level of paging.
 
 **/
 VOID
 SetPageTablePoolReadOnly (
   IN  UINTN                 PageTableBase,
   IN  EFI_PHYSICAL_ADDRESS  Address,
-  IN  BOOLEAN               Level4Paging
+  IN  UINT8                 LevelofPaging
   )
```
Comment (niruiyu) [inline MdeModulePkg/Core/DxeIplPeim/X64/VirtualMemory.c:496]: "LevelOfPaging"

File: UefiPayloadPkg/UefiPayloadEntry/X64/VirtualMemory.c [hunk 2]
```diff
@@ -481,14 +481,14 @@ Split1GPageTo2M (
...
-  @param[in] Level4Paging     Level 4 paging flag.
+  @param[in] LevelofPaging     Level of paging.
 
 **/
 VOID
 SetPageTablePoolReadOnly (
   IN  UINTN                 PageTableBase,
   IN  EFI_PHYSICAL_ADDRESS  Address,
-  IN  BOOLEAN               Level4Paging
+  IN  UINT8                 LevelofPaging
```
Comment (niruiyu) [inline UefiPayloadPkg/UefiPayloadEntry/X64/VirtualMemory.c:491]: LevelOfPaging

**Review by niruiyu — APPROVED** [review]

**Comment by niruiyu** [comment]
> @niruiyu Is this for edk2-stable202411? Please update title if that is the request and add justification.

The patch fixes a critical bug when 5-level paging is enabled in the 64-bit PEI phase. The DxeIpl that runs at the end of PEI and the UplEntry that runs at the beginning of Payload both do not handle 5-l paging properly.

What title should be used for the stable-tag required changes? @NingFengGit can help to change accordingly.

**Comment by mdkinney** [comment]
Approved for edk2-stable202411

**Review by mdkinney — APPROVED** [review]

**Pull request merged**